{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# A pipeline to search for the presence of a listed set of non-synonymous mutaitons in the reads of a given sorted `.bam` file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from collections import Counter\n",
    "import re\n",
    "\n",
    "import ConfigParser"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## I. User inputs\n",
    "\n",
    "Later, this will be: (1) a sorted `.bam` file and (2) an `.ini` file containing mutation information (3) a downsamlping flag."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_bam = \"results/ont/uk/2019-03-20/BC03.pf.bam\"\n",
    "input_path = os.path.dirname(input_bam)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ini = \"data/resources/pf-regions/kelch13.ini\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/resources/pf-regions/kelch13.ini']"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config = ConfigParser.ConfigParser()\n",
    "config.read(gene_ini)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_dt = {}\n",
    "gene_dt[\"genome\"] = config.get(\"Location\", \"genome\")\n",
    "gene_dt[\"chromosome\"] = config.get(\"Location\", \"chromosome\")\n",
    "gene_dt[\"start\"] = config.getint(\"Location\", \"start\")\n",
    "gene_dt[\"end\"] = config.getint(\"Location\", \"end\")\n",
    "gene_dt[\"strand\"] = config.get(\"Location\", \"strand\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [],
   "source": [
    "downsample = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output directory\n",
    "output_path = input_path.replace(\"results\", \"analysis\")\n",
    "if not os.path.isdir(output_path):\n",
    "    os.mkdir(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/ont/uk/2019-03-20\n",
      "analysis/ont/uk/2019-03-20\n"
     ]
    }
   ],
   "source": [
    "print input_path\n",
    "print output_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## II. Execute Downsampling\n",
    "\n",
    "Useful in cases where there is more read data than necessary -- or when looking to speed up analysis (e.g. beta-testing)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downsampling from SAM file...\n",
      "Converting to BAM...\n",
      "Sorting BAM...\n",
      "Indexing BAM...\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "if downsample:\n",
    "    print \"Downsampling from SAM file...\"\n",
    "    n_reads = 1000\n",
    "    \n",
    "    # You have to downsample from the `.sam` file\n",
    "    input_sam = input_bam.replace(\"bam\", \"sam\")\n",
    "    dwn_sam = os.path.join(output_path, os.path.basename(input_sam))\n",
    "    dwn_bam = dwn_sam.replace(\"sam\", \"bam\")\n",
    "    dwn_sorted_bam = dwn_bam.replace(\"bam\", \"sorted.bam\")\n",
    "    \n",
    "    # Downsample by shuffling lines, first extract header\n",
    "    # Get header\n",
    "    os.system(\"grep '^@' %s > %s\" % (input_sam, dwn_sam))\n",
    "\n",
    "    # Downsample\n",
    "    os.system(\"sed '/^@/d' %s > no_header.tmp.bam\" % input_sam)\n",
    "    os.system(\"gshuf -n %d no_header.tmp.bam >> %s\" % (n_reads, dwn_sam))\n",
    "    os.system(\"rm no_header.tmp.bam\")\n",
    "    \n",
    "    # Prepare file for pileup\n",
    "    print \"Converting to BAM...\"\n",
    "    os.system(\"samtools view -S -b %s > %s\" % (dwn_sam, dwn_bam))\n",
    "    print \"Sorting BAM...\"\n",
    "    os.system(\"samtools sort %s -o %s\" % (dwn_sam, dwn_sorted_bam))\n",
    "    print \"Indexing BAM...\"\n",
    "    os.system(\"samtools index %s\" % (dwn_sorted_bam))\n",
    "    print \"Done.\"\n",
    "    \n",
    "    pileup_bam = dwn_sorted_bam\n",
    "else:\n",
    "    pileup_bam = input_bam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## III. Excute `samtools mpileup`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating pileup...\n",
      "samtools mpileup -f data/resources/plasmodb/39/PlasmoDB-39_Pfalciparum3D7_Genome.fasta -r Pf3D7_13_v3:1724817-1726997 -Q 0 -aa -B analysis/ont/uk/2019-03-20/BC03.pf.sorted.bam > analysis/ont/uk/2019-03-20/BC03.pf.pileup\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "pileup_path = pileup_bam.replace(\"sorted.bam\", \"pileup\")\n",
    "position = gene_dt[\"chromosome\"] + \":\" + str(gene_dt[\"start\"]) + \"-\" + str(gene_dt[\"end\"])\n",
    "\n",
    "cmd = \"samtools mpileup -f %s -r %s -Q 0 -aa -B %s > %s\" % (gene_dt[\"genome\"], position, \n",
    "                                                            pileup_bam, pileup_path)\n",
    "print \"Generating pileup...\"\n",
    "print cmd\n",
    "os.system(cmd)\n",
    "print \"Done.\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## IV. Search for Mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene is on reverse strand, inverting pileup.\n",
      "Note: still need to perform reverse complementation.\n"
     ]
    }
   ],
   "source": [
    "# Make necessary prepartions if the gene is on the reverse strand\n",
    "if gene_dt['strand'] == 'reverse':\n",
    "    print \"Gene is on reverse strand, inverting pileup.\"\n",
    "    print \"Note: still need to perform reverse complementation.\"\n",
    "    pileup_path_reverse = pileup_path.replace(\"pileup\", \"reverse.pileup\")\n",
    "    os.system('tail -r %s > %s' % (pileup_path, pileup_path_reverse))\n",
    "    pileup_path = pileup_path_reverse\n",
    "else:\n",
    "    print \"Gene is on forward strand.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations = config.get(\"Mutations\", \"listed\").split(\", \")\n",
    "mutations\n",
    "n_mutations = len(mutations)\n",
    "\n",
    "mutation_dt = {\n",
    "    \"mutation\": [],\n",
    "    \"detected\": [],\n",
    "    \"total_count\": [],\n",
    "    \"major_amino\": [],\n",
    "    \"major_count\": [],\n",
    "    \"mutation_amino\": [],\n",
    "    \"mutation_count\": [],\n",
    "    \"n_aminos\": [],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====================================================================================================\n",
      "Searching for...\n",
      "  Mutation: N458Y\n",
      "  Codon: 458\n",
      "  Corresponding bases: [1371 1372 1373]\n",
      "Discovered...\n",
      "  Reference codon (from 3D7): AAT\n",
      "  Majority codon (from pileup): AAT\n",
      "  Number of unique codons discovered (including indels): 21\n",
      "  Reference codon count: 409\n",
      "  Majority codon count: 409\n",
      "  Total codon count: 495\n",
      "\n",
      "  Reference amino (from 3D7): N\n",
      "  Majority amino (from pileup): N\n",
      "  Number of unique aminos discovered (including indels): 8\n",
      "  Reference amino count: 410\n",
      "  Majority amino count: 410\n",
      "  Total amino count: 428\n",
      "Mutation detected?: True\n",
      "Mutation count: 1\n",
      "Percent of total: 0.23%\n",
      "====================================================================================================\n",
      "====================================================================================================\n",
      "Searching for...\n",
      "  Mutation: Y493H\n",
      "  Codon: 493\n",
      "  Corresponding bases: [1476 1477 1478]\n",
      "Discovered...\n",
      "  Reference codon (from 3D7): TAC\n",
      "  Majority codon (from pileup): TAC\n",
      "  Number of unique codons discovered (including indels): 32\n",
      "  Reference codon count: 339\n",
      "  Majority codon count: 339\n",
      "  Total codon count: 504\n",
      "\n",
      "  Reference amino (from 3D7): Y\n",
      "  Majority amino (from pileup): Y\n",
      "  Number of unique aminos discovered (including indels): 8\n",
      "  Reference amino count: 353\n",
      "  Majority amino count: 353\n",
      "  Total amino count: 403\n",
      "Mutation detected?: True\n",
      "Mutation count: 4\n",
      "Percent of total: 0.99%\n",
      "====================================================================================================\n",
      "====================================================================================================\n",
      "Searching for...\n",
      "  Mutation: R539T\n",
      "  Codon: 539\n",
      "  Corresponding bases: [1614 1615 1616]\n",
      "Discovered...\n",
      "  Reference codon (from 3D7): AGA\n",
      "  Majority codon (from pileup): AGA\n",
      "  Number of unique codons discovered (including indels): 21\n",
      "  Reference codon count: 358\n",
      "  Majority codon count: 358\n",
      "  Total codon count: 521\n",
      "\n",
      "  Reference amino (from 3D7): R\n",
      "  Majority amino (from pileup): R\n",
      "  Number of unique aminos discovered (including indels): 5\n",
      "  Reference amino count: 364\n",
      "  Majority amino count: 364\n",
      "  Total amino count: 382\n",
      "Mutation detected?: True\n",
      "Mutation count: 1\n",
      "Percent of total: 0.26%\n",
      "====================================================================================================\n",
      "====================================================================================================\n",
      "Searching for...\n",
      "  Mutation: I543T\n",
      "  Codon: 543\n",
      "  Corresponding bases: [1626 1627 1628]\n",
      "Discovered...\n",
      "  Reference codon (from 3D7): ATT\n",
      "  Majority codon (from pileup): ATT\n",
      "  Number of unique codons discovered (including indels): 35\n",
      "  Reference codon count: 370\n",
      "  Majority codon count: 370\n",
      "  Total codon count: 500\n",
      "\n",
      "  Reference amino (from 3D7): I\n",
      "  Majority amino (from pileup): I\n",
      "  Number of unique aminos discovered (including indels): 7\n",
      "  Reference amino count: 376\n",
      "  Majority amino count: 376\n",
      "  Total amino count: 396\n",
      "Mutation detected?: True\n",
      "Mutation count: 5\n",
      "Percent of total: 1.26%\n",
      "====================================================================================================\n",
      "====================================================================================================\n",
      "Searching for...\n",
      "  Mutation: R561H\n",
      "  Codon: 561\n",
      "  Corresponding bases: [1680 1681 1682]\n",
      "Discovered...\n",
      "  Reference codon (from 3D7): CGT\n",
      "  Majority codon (from pileup): CGT\n",
      "  Number of unique codons discovered (including indels): 29\n",
      "  Reference codon count: 360\n",
      "  Majority codon count: 360\n",
      "  Total codon count: 500\n",
      "\n",
      "  Reference amino (from 3D7): R\n",
      "  Majority amino (from pileup): R\n",
      "  Number of unique aminos discovered (including indels): 7\n",
      "  Reference amino count: 386\n",
      "  Majority amino count: 386\n",
      "  Total amino count: 417\n",
      "Mutation detected?: True\n",
      "Mutation count: 17\n",
      "Percent of total: 4.08%\n",
      "====================================================================================================\n",
      "====================================================================================================\n",
      "Searching for...\n",
      "  Mutation: C580Y\n",
      "  Codon: 580\n",
      "  Corresponding bases: [1737 1738 1739]\n",
      "Discovered...\n",
      "  Reference codon (from 3D7): TGT\n",
      "  Majority codon (from pileup): TGT\n",
      "  Number of unique codons discovered (including indels): 23\n",
      "  Reference codon count: 411\n",
      "  Majority codon count: 411\n",
      "  Total codon count: 502\n",
      "\n",
      "  Reference amino (from 3D7): C\n",
      "  Majority amino (from pileup): C\n",
      "  Number of unique aminos discovered (including indels): 5\n",
      "  Reference amino count: 412\n",
      "  Majority amino count: 412\n",
      "  Total amino count: 439\n",
      "Mutation detected?: True\n",
      "Mutation count: 10\n",
      "Percent of total: 2.28%\n",
      "====================================================================================================\n",
      "====================================================================================================\n",
      "Searching for...\n",
      "  Mutation: P441L\n",
      "  Codon: 441\n",
      "  Corresponding bases: [1320 1321 1322]\n",
      "Discovered...\n",
      "  Reference codon (from 3D7): CCA\n",
      "  Majority codon (from pileup): CCA\n",
      "  Number of unique codons discovered (including indels): 19\n",
      "  Reference codon count: 423\n",
      "  Majority codon count: 423\n",
      "  Total codon count: 506\n",
      "\n",
      "  Reference amino (from 3D7): P\n",
      "  Majority amino (from pileup): P\n",
      "  Number of unique aminos discovered (including indels): 4\n",
      "  Reference amino count: 429\n",
      "  Majority amino count: 429\n",
      "  Total amino count: 440\n",
      "Mutation detected?: True\n",
      "Mutation count: 5\n",
      "Percent of total: 1.14%\n",
      "====================================================================================================\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "\"['major_amino_count'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-212-88ea85c7692b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    106\u001b[0m                            \u001b[0;34m\"major_amino\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"major_amino_count\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    107\u001b[0m                            \u001b[0;34m\"mutation_amino\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"mutation_count\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 108\u001b[0;31m                            \"n_aminos\"]]\n\u001b[0m",
      "\u001b[0;32m/anaconda2/lib/python2.7/site-packages/pandas/core/frame.pyc\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2680\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mSeries\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndarray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mIndex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2681\u001b[0m             \u001b[0;31m# either boolean or fancy integer index\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2682\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2683\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2684\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_frame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda2/lib/python2.7/site-packages/pandas/core/frame.pyc\u001b[0m in \u001b[0;36m_getitem_array\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2724\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_take\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2725\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2726\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_convert_to_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2727\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_take\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2728\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda2/lib/python2.7/site-packages/pandas/core/indexing.pyc\u001b[0m in \u001b[0;36m_convert_to_indexer\u001b[0;34m(self, obj, axis, is_setter)\u001b[0m\n\u001b[1;32m   1325\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1326\u001b[0m                     raise KeyError('{mask} not in index'\n\u001b[0;32m-> 1327\u001b[0;31m                                    .format(mask=objarr[mask]))\n\u001b[0m\u001b[1;32m   1328\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1329\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values_from_object\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['major_amino_count'] not in index\""
     ]
    }
   ],
   "source": [
    "for mutation in mutations:\n",
    "    \n",
    "    # Parse Mutation Information\n",
    "    print(\"====================================================================================================\")\n",
    "    print(\"Searching for...\")\n",
    "    print(\"  Mutation:\", mutation)\n",
    "    codon = int(mutation[1:-1])\n",
    "    codon_nts = np.arange(3*(codon - 1), 3*codon)\n",
    "    print(\"  Codon:\", codon)\n",
    "    print(\"  Corresponding bases:\", codon_nts)\n",
    "    amino_alt = mutation[-1]\n",
    "    \n",
    "    with open(pileup_path, \"r\") as fn:\n",
    "        \n",
    "        # Extract pileups corresponding to codon position\n",
    "        codon_ref_nts = []\n",
    "        codon_pileup = []\n",
    "        for i, line in enumerate(fn):\n",
    "            if i in codon_nts:\n",
    "                \n",
    "                # Extract pileup information\n",
    "                chrom, pos, ref, coverage, pileup, _ = line.split(\"\\t\")\n",
    "                \n",
    "                # Convert the pileup string to A, T, C, G, +, -\n",
    "                processed_pileup = process_pileup(pileup, ref)\n",
    "                \n",
    "                # Reverse complement if necessary\n",
    "                if gene_dt[\"strand\"] == 'reverse':\n",
    "                    complement_map = {\"A\": \"T\", \"T\": \"A\", \n",
    "                                      \"G\": \"C\", \"C\": \"G\", \n",
    "                                      \"-\": \"-\", \"+\": \"+\" }\n",
    "                    ref = complement_map[ref]\n",
    "                    processed_pileup = \"\".join([complement_map[base] for base in processed_pileup])\n",
    "                \n",
    "                # Append \n",
    "                codon_ref_nts.append(ref)\n",
    "                codon_pileup.append(processed_pileup)\n",
    "                \n",
    "        # Check complete pileup has been extracted\n",
    "        assert len(codon_ref_nts) == 3\n",
    "        assert len(codon_pileup) == 3\n",
    "        \n",
    "        # Pre-process from lists into strings\n",
    "        codon_ref = \"\".join(codon_ref_nts)\n",
    "        codon_pileup = [\"\".join(c) for c in zip(*codon_pileup)]\n",
    "        \n",
    "        # Get frequencies of codons (i.e. nucleotide level)\n",
    "        codon_frequencies = Counter(codon_pileup)\n",
    "        major_codon, major_codon_count = codon_frequencies.most_common(1)[0]\n",
    "        ref_codon_count = codon_frequencies[codon_ref]\n",
    "        total_codon_count = sum(codon_frequencies.values())\n",
    "        \n",
    "        print(\"Discovered...\")\n",
    "        print(\"  Reference codon (from 3D7):\", codon_ref)\n",
    "        print(\"  Majority codon (from pileup):\", major_codon)\n",
    "        print(\"  Number of unique codons discovered (including indels):\", len(codon_frequencies))\n",
    "        print(\"  Reference codon count:\", ref_codon_count)\n",
    "        print(\"  Majority codon count:\", major_codon_count)\n",
    "        print(\"  Total codon count:\", total_codon_count)\n",
    "        print(\"\")\n",
    "        \n",
    "        # Get frequencies of amino acids\n",
    "        amino_ref = codon_to_amino(codon_ref, genetic_code)\n",
    "        amino_frequencies = Counter([codon_to_amino(c, genetic_code) for c in codon_pileup])\n",
    "        # next line removes indels which have prevented making amino acid calls\n",
    "        amino_frequencies = Counter(dict([(k, v) for k, v in amino_frequencies.iteritems() if k != None]))\n",
    "        major_amino, major_amino_count = amino_frequencies.most_common(1)[0]\n",
    "        ref_amino_count = amino_frequencies[amino_ref]\n",
    "        total_amino_count = sum(amino_frequencies.values())\n",
    "        \n",
    "        print(\"  Reference amino (from 3D7):\", amino_ref)\n",
    "        print(\"  Majority amino (from pileup):\", major_amino)\n",
    "        print(\"  Number of unique aminos discovered (including indels):\", len(amino_frequencies))\n",
    "        print(\"  Reference amino count:\", ref_amino_count)\n",
    "        print(\"  Majority amino count:\", major_amino_count)\n",
    "        print(\"  Total amino count:\", total_amino_count)\n",
    "            \n",
    "        # Finally, check for non-synonymous change of interest\n",
    "        if amino_alt in amino_frequencies.keys():\n",
    "            mutation_detected = True\n",
    "            mutation_count = amino_frequencies[amino_alt]\n",
    "        else:\n",
    "            mutation_detected = False\n",
    "            mutation_count = 0\n",
    "            \n",
    "        print(\"Mutation detected?:\", mutation_detected)\n",
    "        print(\"Mutation count:\", mutation_count)\n",
    "        print(\"Percent of total: %.02f%%\" % (100*float(mutation_count)/total_amino_count))\n",
    "        \n",
    "        \n",
    "        # Store output\n",
    "        mutation_dt[\"mutation\"].append(mutation)\n",
    "        mutation_dt[\"detected\"].append(mutation_detected)\n",
    "        mutation_dt[\"total_count\"].append(total_amino_count)\n",
    "        mutation_dt[\"major_amino\"].append(major_amino)\n",
    "        mutation_dt[\"major_count\"].append(major_amino_count)\n",
    "        mutation_dt[\"mutation_amino\"].append(amino_alt)\n",
    "        mutation_dt[\"mutation_count\"].append(mutation_count)\n",
    "        mutation_dt[\"n_aminos\"].append(len(amino_frequencies))\n",
    "        \n",
    "        print(\"====================================================================================================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_df = pd.DataFrame(mutation_dt)\n",
    "mutation_df = mutation_df[[\"mutation\", \"detected\", \n",
    "                           \"total_count\",\n",
    "                           \"major_amino\", \"major_count\",\n",
    "                           \"mutation_amino\", \"mutation_count\",\n",
    "                           \"n_aminos\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation</th>\n",
       "      <th>detected</th>\n",
       "      <th>total_count</th>\n",
       "      <th>major_amino</th>\n",
       "      <th>major_count</th>\n",
       "      <th>mutation_amino</th>\n",
       "      <th>mutation_count</th>\n",
       "      <th>n_aminos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N458Y</td>\n",
       "      <td>True</td>\n",
       "      <td>428</td>\n",
       "      <td>N</td>\n",
       "      <td>410</td>\n",
       "      <td>Y</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y493H</td>\n",
       "      <td>True</td>\n",
       "      <td>403</td>\n",
       "      <td>Y</td>\n",
       "      <td>353</td>\n",
       "      <td>H</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R539T</td>\n",
       "      <td>True</td>\n",
       "      <td>382</td>\n",
       "      <td>R</td>\n",
       "      <td>364</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I543T</td>\n",
       "      <td>True</td>\n",
       "      <td>396</td>\n",
       "      <td>I</td>\n",
       "      <td>376</td>\n",
       "      <td>T</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R561H</td>\n",
       "      <td>True</td>\n",
       "      <td>417</td>\n",
       "      <td>R</td>\n",
       "      <td>386</td>\n",
       "      <td>H</td>\n",
       "      <td>17</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C580Y</td>\n",
       "      <td>True</td>\n",
       "      <td>439</td>\n",
       "      <td>C</td>\n",
       "      <td>412</td>\n",
       "      <td>Y</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P441L</td>\n",
       "      <td>True</td>\n",
       "      <td>440</td>\n",
       "      <td>P</td>\n",
       "      <td>429</td>\n",
       "      <td>L</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  mutation  detected  total_count major_amino  major_count mutation_amino  \\\n",
       "0    N458Y      True          428           N          410              Y   \n",
       "1    Y493H      True          403           Y          353              H   \n",
       "2    R539T      True          382           R          364              T   \n",
       "3    I543T      True          396           I          376              T   \n",
       "4    R561H      True          417           R          386              H   \n",
       "5    C580Y      True          439           C          412              Y   \n",
       "6    P441L      True          440           P          429              L   \n",
       "\n",
       "   mutation_count  n_aminos  \n",
       "0               1         8  \n",
       "1               4         8  \n",
       "2               1         5  \n",
       "3               5         7  \n",
       "4              17         7  \n",
       "5              10         5  \n",
       "6               5         4  "
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## -. Library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_pileup(pileup, ref):\n",
    "    \"\"\"\n",
    "    Process a read pileup from \n",
    "    `samtools mpileup` into\n",
    "    A,T,C,G,+,-\n",
    "    \n",
    "    Note, the solution is ugly, but\n",
    "    re.sub(\\+([0-9])[ATCGatcg]+, ... fails when a variant\n",
    "    follows directly after an insertion deltion.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    n_pileup = len(pileup)\n",
    "    processed_pileup = ''\n",
    "    i = 0\n",
    "    while i < n_pileup:\n",
    "        p = pileup[i]\n",
    "        if p in [\"+\", \"-\"]:\n",
    "            size = int(pileup[i + 1])\n",
    "            j = size + 2\n",
    "            processed_pileup += p\n",
    "        elif p == \"*\":\n",
    "            j = 1\n",
    "            processed_pileup += \"-\"\n",
    "        else:\n",
    "            j = 1\n",
    "            processed_pileup += p\n",
    "        i += j\n",
    "    \n",
    "    processed_pileup = re.sub(\"\\$|\\^\\]\", \"\", processed_pileup)\n",
    "    processed_pileup = re.sub(\"\\.|,\", ref, processed_pileup)\n",
    "    processed_pileup = processed_pileup.upper()\n",
    "    \n",
    "    return processed_pileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [],
   "source": [
    "genetic_code = { \n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', \n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T', \n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', \n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                  \n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', \n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P', \n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', \n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R', \n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', \n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A', \n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', \n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G', \n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', \n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L', \n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_', \n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W', \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codon_to_amino(codon, genetic_code):\n",
    "    if codon in genetic_code.keys():\n",
    "        amino = genetic_code[codon]\n",
    "    else:\n",
    "        amino = None\n",
    "    return amino"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
