{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MMPP: Mobile Malaria Project Pipelines\n",
    "# ----------------------------------------\n",
    "# Search for the presence of \n",
    "# a listed set of non-synonymous mutations\n",
    "# in a specified gene \n",
    "# amongst the reads of a sorted `.bam` file\n",
    "# ----------------------------------------\n",
    "# JHendry, 2019/03/27"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import configparser\n",
    "import getopt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from lib.mutation import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User Inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_bam = \"data/ont/uk/2019-03-28/BC08.sorted.bam\"\n",
    "gene_ini = \"data/resources/pf-regions/kelch13.ini\"\n",
    "min_freq = 0.1\n",
    "downsample = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path = os.path.dirname(input_bam)\n",
    "output_path = input_path.replace(\"data\", \"analysis\")\n",
    "if not os.path.isdir(output_path):\n",
    "    os.mkdir(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = configparser.ConfigParser()\n",
    "config.read(gene_ini)\n",
    "\n",
    "# hold gene location information\n",
    "gene_dt = {}\n",
    "gene_dt[\"name\"] = config.get(\"Location\", \"name\")\n",
    "gene_dt[\"genome\"] = config.get(\"Location\", \"genome\")\n",
    "gene_dt[\"chromosome\"] = config.get(\"Location\", \"chromosome\")\n",
    "gene_dt[\"start\"] = config.getint(\"Location\", \"start\")\n",
    "gene_dt[\"end\"] = config.getint(\"Location\", \"end\")\n",
    "gene_dt[\"strand\"] = config.get(\"Location\", \"strand\")\n",
    "     \n",
    "# hold mutation information\n",
    "mutations = config.get(\"Mutations\", \"listed\").split(\", \")\n",
    "n_mutations = len(mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================================\n",
      "MMP Mutation Search Pipeline\n",
      "--------------------------------------------------------------------------------\n",
      "Gene: KELCH13\n",
      "Chromosome: Pf3D7_13_v3\n",
      "Start: 1724817\n",
      "End: 1726997\n",
      "Strand: reverse\n",
      "\n",
      "Searching for 31 mutations.\n",
      "\n",
      "Input BAM: data/ont/uk/2019-03-28/BC08.sorted.bam\n",
      "Output path: analysis/ont/uk/2019-03-28\n",
      "Reference genome: data/resources/plasmodb/39/PlasmoDB-39_Pfalciparum3D7_Genome.fasta\n",
      "\n",
      "Downsampling? False\n",
      "================================================================================\n"
     ]
    }
   ],
   "source": [
    "print(\"================================================================================\")\n",
    "print(\"MMP Mutation Search Pipeline\")\n",
    "print(\"--------------------------------------------------------------------------------\")\n",
    "print(\"Gene:\", gene_dt[\"name\"])\n",
    "print(\"Chromosome:\", gene_dt[\"chromosome\"])\n",
    "print(\"Start:\", gene_dt[\"start\"])\n",
    "print(\"End:\", gene_dt[\"end\"])\n",
    "print(\"Strand:\", gene_dt[\"strand\"])\n",
    "print(\"\")\n",
    "print(\"Searching for %d mutations.\" % n_mutations)\n",
    "print(\"\")\n",
    "print(\"Input BAM:\", input_bam)\n",
    "print(\"Output path:\", output_path)\n",
    "print(\"Reference genome:\", gene_dt[\"genome\"])\n",
    "print(\"\")\n",
    "print(\"Downsampling?\", downsample)\n",
    "print(\"================================================================================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "pileup_bam = input_bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generating pileup...\n",
      "samtools mpileup -f data/resources/plasmodb/39/PlasmoDB-39_Pfalciparum3D7_Genome.fasta -r Pf3D7_13_v3:1724817-1726997 -Q 0 -aa -B data/ont/uk/2019-03-28/BC08.sorted.bam > data/ont/uk/2019-03-28/BC08.pileup\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "# Generate read pileup using samtools\n",
    "pileup_path = pileup_bam.replace(\"sorted.bam\", \"pileup\")\n",
    "position = gene_dt[\"chromosome\"] + \":\" + str(gene_dt[\"start\"]) + \"-\" + str(gene_dt[\"end\"])\n",
    "\n",
    "cmd = \"samtools mpileup -f %s -r %s -Q 0 -aa -B %s > %s\" % (gene_dt[\"genome\"], \n",
    "                                                            position, \n",
    "                                                            pileup_bam, \n",
    "                                                            pileup_path)\n",
    "print(\"Generating pileup...\")\n",
    "print(cmd)\n",
    "os.system(cmd)\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene is on reverse strand, inverting pileup.\n",
      "Note: still need to perform reverse complementation.\n"
     ]
    }
   ],
   "source": [
    "# Make necessary prepartions if the gene is on the reverse strand\n",
    "if gene_dt['strand'] == 'reverse':\n",
    "    print(\"Gene is on reverse strand, inverting pileup.\")\n",
    "    print(\"Note: still need to perform reverse complementation.\")\n",
    "    pileup_path_reverse = pileup_path.replace(\"pileup\", \"reverse.pileup\")\n",
    "    os.system('tail -r %s > %s' % (pileup_path, pileup_path_reverse))\n",
    "    pileup_path = pileup_path_reverse\n",
    "else:\n",
    "    print(\"Gene is on forward strand.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Search for mutations\n",
    "mutation_dt = {\n",
    "    \"position\": [],\n",
    "    \"total_counts\": [],\n",
    "    \"total_counts_noindel\": [],\n",
    "    \"ref_codon\": [],\n",
    "    \"ref_amino\": [],\n",
    "    \"ref_freq\": [],\n",
    "    \"major_codon\": [],\n",
    "    \"major_amino\": [],\n",
    "    \"major_freq\": [],\n",
    "    \"n_mutation_types\": [],\n",
    "    \"mutation_codon\": [],\n",
    "    \"mutation_amino\": [],\n",
    "    \"mutation_freq\": [],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 1\n",
      "Discovered...\n",
      "  Total codon count (including indels): 373\n",
      "  Reference codon (from 3D7): ATG\n",
      "  Reference codon count: 291\n",
      "  Majority codon (from pileup): ATG\n",
      "  Majority codon count: 291\n",
      "  Number of unique codons discovered (including indels): 18\n",
      "\n",
      "  Total codon count (excluding indels): 304\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: ACG = T\n",
      "                        at frequency: 0.00987\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 2\n",
      "Discovered...\n",
      "  Total codon count (including indels): 373\n",
      "  Reference codon (from 3D7): GAA\n",
      "  Reference codon count: 308\n",
      "  Majority codon (from pileup): GAA\n",
      "  Majority codon count: 308\n",
      "  Number of unique codons discovered (including indels): 15\n",
      "\n",
      "  Total codon count (excluding indels): 312\n",
      "    Unique types: 5\n",
      "    Non-synonymous: 3\n",
      "\n",
      "    Highest frequency non-synonymous: GAT = D\n",
      "                        at frequency: 0.00321\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 3\n",
      "Discovered...\n",
      "  Total codon count (including indels): 375\n",
      "  Reference codon (from 3D7): GGA\n",
      "  Reference codon count: 313\n",
      "  Majority codon (from pileup): GGA\n",
      "  Majority codon count: 313\n",
      "  Number of unique codons discovered (including indels): 12\n",
      "\n",
      "  Total codon count (excluding indels): 313\n",
      "    Unique types: 1\n",
      "    Non-synonymous: 0\n",
      "\n",
      "    No non-synonymous mutations discovered.\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 4\n",
      "Discovered...\n",
      "  Total codon count (including indels): 386\n",
      "  Reference codon (from 3D7): GAA\n",
      "  Reference codon count: 258\n",
      "  Majority codon (from pileup): GAA\n",
      "  Majority codon count: 258\n",
      "  Number of unique codons discovered (including indels): 24\n",
      "\n",
      "  Total codon count (excluding indels): 276\n",
      "    Unique types: 9\n",
      "    Non-synonymous: 7\n",
      "\n",
      "    Highest frequency non-synonymous: GCA = A\n",
      "                        at frequency: 0.01449\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 5\n",
      "Discovered...\n",
      "  Total codon count (including indels): 377\n",
      "  Reference codon (from 3D7): AAA\n",
      "  Reference codon count: 146\n",
      "  Majority codon (from pileup): AA-\n",
      "  Majority codon count: 148\n",
      "  Number of unique codons discovered (including indels): 22\n",
      "\n",
      "  Total codon count (excluding indels): 152\n",
      "    Unique types: 5\n",
      "    Non-synonymous: 3\n",
      "\n",
      "    Highest frequency non-synonymous: AGA = R\n",
      "                        at frequency: 0.01974\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 6\n",
      "Discovered...\n",
      "  Total codon count (including indels): 380\n",
      "  Reference codon (from 3D7): GTA\n",
      "  Reference codon count: 179\n",
      "  Majority codon (from pileup): GTA\n",
      "  Majority codon count: 179\n",
      "  Number of unique codons discovered (including indels): 24\n",
      "\n",
      "  Total codon count (excluding indels): 194\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 4\n",
      "\n",
      "    Highest frequency non-synonymous: GCA = A\n",
      "                        at frequency: 0.03093\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 7\n",
      "Discovered...\n",
      "  Total codon count (including indels): 374\n",
      "  Reference codon (from 3D7): AAA\n",
      "  Reference codon count: 321\n",
      "  Majority codon (from pileup): AAA\n",
      "  Majority codon count: 321\n",
      "  Number of unique codons discovered (including indels): 18\n",
      "\n",
      "  Total codon count (excluding indels): 331\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: GAA = E\n",
      "                        at frequency: 0.00906\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 8\n",
      "Discovered...\n",
      "  Total codon count (including indels): 378\n",
      "  Reference codon (from 3D7): ACA\n",
      "  Reference codon count: 133\n",
      "  Majority codon (from pileup): ACA\n",
      "  Majority codon count: 133\n",
      "  Number of unique codons discovered (including indels): 27\n",
      "\n",
      "  Total codon count (excluding indels): 167\n",
      "    Unique types: 10\n",
      "    Non-synonymous: 7\n",
      "\n",
      "    Highest frequency non-synonymous: CCA = P\n",
      "                        at frequency: 0.06587\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 9\n",
      "Discovered...\n",
      "  Total codon count (including indels): 374\n",
      "  Reference codon (from 3D7): AAA\n",
      "  Reference codon count: 290\n",
      "  Majority codon (from pileup): AAA\n",
      "  Majority codon count: 290\n",
      "  Number of unique codons discovered (including indels): 18\n",
      "\n",
      "  Total codon count (excluding indels): 299\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 5\n",
      "\n",
      "    Highest frequency non-synonymous: AAT = N\n",
      "                        at frequency: 0.00669\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 10\n",
      "Discovered...\n",
      "  Total codon count (including indels): 383\n",
      "  Reference codon (from 3D7): GCA\n",
      "  Reference codon count: 235\n",
      "  Majority codon (from pileup): GCA\n",
      "  Majority codon count: 235\n",
      "  Number of unique codons discovered (including indels): 30\n",
      "\n",
      "  Total codon count (excluding indels): 259\n",
      "    Unique types: 10\n",
      "    Non-synonymous: 8\n",
      "\n",
      "    Highest frequency non-synonymous: ACA = T\n",
      "                        at frequency: 0.04633\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 11\n",
      "Discovered...\n",
      "  Total codon count (including indels): 374\n",
      "  Reference codon (from 3D7): AAT\n",
      "  Reference codon count: 278\n",
      "  Majority codon (from pileup): AAT\n",
      "  Majority codon count: 278\n",
      "  Number of unique codons discovered (including indels): 22\n",
      "\n",
      "  Total codon count (excluding indels): 295\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: ACT = T\n",
      "                        at frequency: 0.01017\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 12\n",
      "Discovered...\n",
      "  Total codon count (including indels): 381\n",
      "  Reference codon (from 3D7): AGT\n",
      "  Reference codon count: 276\n",
      "  Majority codon (from pileup): AGT\n",
      "  Majority codon count: 276\n",
      "  Number of unique codons discovered (including indels): 24\n",
      "\n",
      "  Total codon count (excluding indels): 311\n",
      "    Unique types: 10\n",
      "    Non-synonymous: 8\n",
      "\n",
      "    Highest frequency non-synonymous: AAT = N\n",
      "                        at frequency: 0.04502\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 13\n",
      "Discovered...\n",
      "  Total codon count (including indels): 381\n",
      "  Reference codon (from 3D7): ATC\n",
      "  Reference codon count: 279\n",
      "  Majority codon (from pileup): ATC\n",
      "  Majority codon count: 279\n",
      "  Number of unique codons discovered (including indels): 22\n",
      "\n",
      "  Total codon count (excluding indels): 307\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 5\n",
      "\n",
      "    Highest frequency non-synonymous: AAC = N\n",
      "                        at frequency: 0.04886\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 14\n",
      "Discovered...\n",
      "  Total codon count (including indels): 371\n",
      "  Reference codon (from 3D7): TCG\n",
      "  Reference codon count: 202\n",
      "  Majority codon (from pileup): TCG\n",
      "  Majority codon count: 202\n",
      "  Number of unique codons discovered (including indels): 30\n",
      "\n",
      "  Total codon count (excluding indels): 230\n",
      "    Unique types: 10\n",
      "    Non-synonymous: 7\n",
      "\n",
      "    Highest frequency non-synonymous: TTG = L\n",
      "                        at frequency: 0.00870\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 15\n",
      "Discovered...\n",
      "  Total codon count (including indels): 375\n",
      "  Reference codon (from 3D7): AAT\n",
      "  Reference codon count: 237\n",
      "  Majority codon (from pileup): AAT\n",
      "  Majority codon count: 237\n",
      "  Number of unique codons discovered (including indels): 29\n",
      "\n",
      "  Total codon count (excluding indels): 285\n",
      "    Unique types: 9\n",
      "    Non-synonymous: 7\n",
      "\n",
      "    Highest frequency non-synonymous: GAT = D\n",
      "                        at frequency: 0.05263\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 16\n",
      "Discovered...\n",
      "  Total codon count (including indels): 369\n",
      "  Reference codon (from 3D7): TTT\n",
      "  Reference codon count: 306\n",
      "  Majority codon (from pileup): TTT\n",
      "  Majority codon count: 306\n",
      "  Number of unique codons discovered (including indels): 20\n",
      "\n",
      "  Total codon count (excluding indels): 325\n",
      "    Unique types: 10\n",
      "    Non-synonymous: 9\n",
      "\n",
      "    Highest frequency non-synonymous: CTT = L\n",
      "                        at frequency: 0.01846\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 17\n",
      "Discovered...\n",
      "  Total codon count (including indels): 369\n",
      "  Reference codon (from 3D7): TCT\n",
      "  Reference codon count: 150\n",
      "  Majority codon (from pileup): TCT\n",
      "  Majority codon count: 150\n",
      "  Number of unique codons discovered (including indels): 23\n",
      "\n",
      "  Total codon count (excluding indels): 164\n",
      "    Unique types: 5\n",
      "    Non-synonymous: 3\n",
      "\n",
      "    Highest frequency non-synonymous: CCT = P\n",
      "                        at frequency: 0.03049\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 18\n",
      "Discovered...\n",
      "  Total codon count (including indels): 390\n",
      "  Reference codon (from 3D7): ATG\n",
      "  Reference codon count: 268\n",
      "  Majority codon (from pileup): ATG\n",
      "  Majority codon count: 268\n",
      "  Number of unique codons discovered (including indels): 21\n",
      "\n",
      "  Total codon count (excluding indels): 288\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: AAG = K\n",
      "                        at frequency: 0.02778\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 19\n",
      "Discovered...\n",
      "  Total codon count (including indels): 372\n",
      "  Reference codon (from 3D7): ACG\n",
      "  Reference codon count: 309\n",
      "  Majority codon (from pileup): ACG\n",
      "  Majority codon count: 309\n",
      "  Number of unique codons discovered (including indels): 18\n",
      "\n",
      "  Total codon count (excluding indels): 322\n",
      "    Unique types: 6\n",
      "    Non-synonymous: 3\n",
      "\n",
      "    Highest frequency non-synonymous: AGG = R\n",
      "                        at frequency: 0.00621\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 20\n",
      "Discovered...\n",
      "  Total codon count (including indels): 375\n",
      "  Reference codon (from 3D7): TAT\n",
      "  Reference codon count: 271\n",
      "  Majority codon (from pileup): TAT\n",
      "  Majority codon count: 271\n",
      "  Number of unique codons discovered (including indels): 23\n",
      "\n",
      "  Total codon count (excluding indels): 297\n",
      "    Unique types: 9\n",
      "    Non-synonymous: 7\n",
      "\n",
      "    Highest frequency non-synonymous: TAA = _\n",
      "                        at frequency: 0.02694\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 21\n",
      "Discovered...\n",
      "  Total codon count (including indels): 370\n",
      "  Reference codon (from 3D7): GAT\n",
      "  Reference codon count: 309\n",
      "  Majority codon (from pileup): GAT\n",
      "  Majority codon count: 309\n",
      "  Number of unique codons discovered (including indels): 20\n",
      "\n",
      "  Total codon count (excluding indels): 320\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 7\n",
      "\n",
      "    Highest frequency non-synonymous: GGT = G\n",
      "                        at frequency: 0.01250\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 22\n",
      "Discovered...\n",
      "  Total codon count (including indels): 374\n",
      "  Reference codon (from 3D7): AGG\n",
      "  Reference codon count: 232\n",
      "  Majority codon (from pileup): AGG\n",
      "  Majority codon count: 232\n",
      "  Number of unique codons discovered (including indels): 38\n",
      "\n",
      "  Total codon count (excluding indels): 283\n",
      "    Unique types: 14\n",
      "    Non-synonymous: 12\n",
      "\n",
      "    Highest frequency non-synonymous: AAG = K\n",
      "                        at frequency: 0.03534\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 23\n",
      "Discovered...\n",
      "  Total codon count (including indels): 373\n",
      "  Reference codon (from 3D7): GAA\n",
      "  Reference codon count: 149\n",
      "  Majority codon (from pileup): GAA\n",
      "  Majority codon count: 149\n",
      "  Number of unique codons discovered (including indels): 26\n",
      "\n",
      "  Total codon count (excluding indels): 158\n",
      "    Unique types: 5\n",
      "    Non-synonymous: 3\n",
      "\n",
      "    Highest frequency non-synonymous: AAA = K\n",
      "                        at frequency: 0.03797\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 24\n",
      "Discovered...\n",
      "  Total codon count (including indels): 378\n",
      "  Reference codon (from 3D7): TCT\n",
      "  Reference codon count: 297\n",
      "  Majority codon (from pileup): TCT\n",
      "  Majority codon count: 297\n",
      "  Number of unique codons discovered (including indels): 22\n",
      "\n",
      "  Total codon count (excluding indels): 315\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 4\n",
      "\n",
      "    Highest frequency non-synonymous: TAT = Y\n",
      "                        at frequency: 0.00635\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 25\n",
      "Discovered...\n",
      "  Total codon count (including indels): 371\n",
      "  Reference codon (from 3D7): GGT\n",
      "  Reference codon count: 205\n",
      "  Majority codon (from pileup): GGT\n",
      "  Majority codon count: 205\n",
      "  Number of unique codons discovered (including indels): 27\n",
      "\n",
      "  Total codon count (excluding indels): 239\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: AGT = S\n",
      "                        at frequency: 0.05021\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 26\n",
      "Discovered...\n",
      "  Total codon count (including indels): 374\n",
      "  Reference codon (from 3D7): GGT\n",
      "  Reference codon count: 225\n",
      "  Majority codon (from pileup): GGT\n",
      "  Majority codon count: 225\n",
      "  Number of unique codons discovered (including indels): 27\n",
      "\n",
      "  Total codon count (excluding indels): 240\n",
      "    Unique types: 6\n",
      "    Non-synonymous: 4\n",
      "\n",
      "    Highest frequency non-synonymous: AGT = S\n",
      "                        at frequency: 0.02083\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 27\n",
      "Discovered...\n",
      "  Total codon count (including indels): 372\n",
      "  Reference codon (from 3D7): AAC\n",
      "  Reference codon count: 236\n",
      "  Majority codon (from pileup): AAC\n",
      "  Majority codon count: 236\n",
      "  Number of unique codons discovered (including indels): 27\n",
      "\n",
      "  Total codon count (excluding indels): 266\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: AAA = K\n",
      "                        at frequency: 0.03008\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 28\n",
      "Discovered...\n",
      "  Total codon count (including indels): 373\n",
      "  Reference codon (from 3D7): AGC\n",
      "  Reference codon count: 241\n",
      "  Majority codon (from pileup): AGC\n",
      "  Majority codon count: 241\n",
      "  Number of unique codons discovered (including indels): 26\n",
      "\n",
      "  Total codon count (excluding indels): 288\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: AAC = N\n",
      "                        at frequency: 0.06250\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 29\n",
      "Discovered...\n",
      "  Total codon count (including indels): 378\n",
      "  Reference codon (from 3D7): AAT\n",
      "  Reference codon count: 285\n",
      "  Majority codon (from pileup): AAT\n",
      "  Majority codon count: 285\n",
      "  Number of unique codons discovered (including indels): 25\n",
      "\n",
      "  Total codon count (excluding indels): 299\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: ACT = T\n",
      "                        at frequency: 0.01003\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 30\n",
      "Discovered...\n",
      "  Total codon count (including indels): 379\n",
      "  Reference codon (from 3D7): AGT\n",
      "  Reference codon count: 275\n",
      "  Majority codon (from pileup): AGT\n",
      "  Majority codon count: 275\n",
      "  Number of unique codons discovered (including indels): 22\n",
      "\n",
      "  Total codon count (excluding indels): 302\n",
      "    Unique types: 10\n",
      "    Non-synonymous: 8\n",
      "\n",
      "    Highest frequency non-synonymous: AAT = N\n",
      "                        at frequency: 0.01987\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 31\n",
      "Discovered...\n",
      "  Total codon count (including indels): 372\n",
      "  Reference codon (from 3D7): GAT\n",
      "  Reference codon count: 304\n",
      "  Majority codon (from pileup): GAT\n",
      "  Majority codon count: 304\n",
      "  Number of unique codons discovered (including indels): 18\n",
      "\n",
      "  Total codon count (excluding indels): 319\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: AAT = N\n",
      "                        at frequency: 0.01254\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 32\n",
      "Discovered...\n",
      "  Total codon count (including indels): 372\n",
      "  Reference codon (from 3D7): GAT\n",
      "  Reference codon count: 283\n",
      "  Majority codon (from pileup): GAT\n",
      "  Majority codon count: 283\n",
      "  Number of unique codons discovered (including indels): 19\n",
      "\n",
      "  Total codon count (excluding indels): 295\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 5\n",
      "\n",
      "    Highest frequency non-synonymous: GTT = V\n",
      "                        at frequency: 0.01017\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 33\n",
      "Discovered...\n",
      "  Total codon count (including indels): 373\n",
      "  Reference codon (from 3D7): AAA\n",
      "  Reference codon count: 280\n",
      "  Majority codon (from pileup): AAA\n",
      "  Majority codon count: 280\n",
      "  Number of unique codons discovered (including indels): 19\n",
      "\n",
      "  Total codon count (excluding indels): 293\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: AAC = N\n",
      "                        at frequency: 0.00683\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 34\n",
      "Discovered...\n",
      "  Total codon count (including indels): 369\n",
      "  Reference codon (from 3D7): AGC\n",
      "  Reference codon count: 171\n",
      "  Majority codon (from pileup): AGC\n",
      "  Majority codon count: 171\n",
      "  Number of unique codons discovered (including indels): 30\n",
      "\n",
      "  Total codon count (excluding indels): 210\n",
      "    Unique types: 6\n",
      "    Non-synonymous: 4\n",
      "\n",
      "    Highest frequency non-synonymous: AAC = N\n",
      "                        at frequency: 0.01429\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 35\n",
      "Discovered...\n",
      "  Total codon count (including indels): 380\n",
      "  Reference codon (from 3D7): GGA\n",
      "  Reference codon count: 169\n",
      "  Majority codon (from pileup): GGA\n",
      "  Majority codon count: 169\n",
      "  Number of unique codons discovered (including indels): 33\n",
      "\n",
      "  Total codon count (excluding indels): 186\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 6\n",
      "\n",
      "    Highest frequency non-synonymous: AGA = R\n",
      "                        at frequency: 0.03226\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 36\n",
      "Discovered...\n",
      "  Total codon count (including indels): 384\n",
      "  Reference codon (from 3D7): AGT\n",
      "  Reference codon count: 269\n",
      "  Majority codon (from pileup): AGT\n",
      "  Majority codon count: 269\n",
      "  Number of unique codons discovered (including indels): 28\n",
      "\n",
      "  Total codon count (excluding indels): 286\n",
      "    Unique types: 10\n",
      "    Non-synonymous: 8\n",
      "\n",
      "    Highest frequency non-synonymous: AGA = R\n",
      "                        at frequency: 0.01049\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 37\n",
      "Discovered...\n",
      "  Total codon count (including indels): 373\n",
      "  Reference codon (from 3D7): AGT\n",
      "  Reference codon count: 293\n",
      "  Majority codon (from pileup): AGT\n",
      "  Majority codon count: 293\n",
      "  Number of unique codons discovered (including indels): 26\n",
      "\n",
      "  Total codon count (excluding indels): 326\n",
      "    Unique types: 9\n",
      "    Non-synonymous: 7\n",
      "\n",
      "    Highest frequency non-synonymous: AGA = R\n",
      "                        at frequency: 0.04294\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 38\n",
      "Discovered...\n",
      "  Total codon count (including indels): 376\n",
      "  Reference codon (from 3D7): AGC\n",
      "  Reference codon count: 246\n",
      "  Majority codon (from pileup): AGC\n",
      "  Majority codon count: 246\n",
      "  Number of unique codons discovered (including indels): 31\n",
      "\n",
      "  Total codon count (excluding indels): 279\n",
      "    Unique types: 10\n",
      "    Non-synonymous: 8\n",
      "\n",
      "    Highest frequency non-synonymous: ATC = I\n",
      "                        at frequency: 0.01792\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 39\n",
      "Discovered...\n",
      "  Total codon count (including indels): 377\n",
      "  Reference codon (from 3D7): GAG\n",
      "  Reference codon count: 222\n",
      "  Majority codon (from pileup): GAG\n",
      "  Majority codon count: 222\n",
      "  Number of unique codons discovered (including indels): 21\n",
      "\n",
      "  Total codon count (excluding indels): 232\n",
      "    Unique types: 5\n",
      "    Non-synonymous: 3\n",
      "\n",
      "    Highest frequency non-synonymous: GGG = G\n",
      "                        at frequency: 0.02155\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 40\n",
      "Discovered...\n",
      "  Total codon count (including indels): 384\n",
      "  Reference codon (from 3D7): AAT\n",
      "  Reference codon count: 260\n",
      "  Majority codon (from pileup): AAT\n",
      "  Majority codon count: 260\n",
      "  Number of unique codons discovered (including indels): 32\n",
      "\n",
      "  Total codon count (excluding indels): 281\n",
      "    Unique types: 12\n",
      "    Non-synonymous: 10\n",
      "\n",
      "    Highest frequency non-synonymous: AGT = S\n",
      "                        at frequency: 0.01423\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 41\n",
      "Discovered...\n",
      "  Total codon count (including indels): 375\n",
      "  Reference codon (from 3D7): GAT\n",
      "  Reference codon count: 281\n",
      "  Majority codon (from pileup): GAT\n",
      "  Majority codon count: 281\n",
      "  Number of unique codons discovered (including indels): 25\n",
      "\n",
      "  Total codon count (excluding indels): 314\n",
      "    Unique types: 9\n",
      "    Non-synonymous: 7\n",
      "\n",
      "    Highest frequency non-synonymous: AAT = N\n",
      "                        at frequency: 0.01274\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 42\n",
      "Discovered...\n",
      "  Total codon count (including indels): 369\n",
      "  Reference codon (from 3D7): TCT\n",
      "  Reference codon count: 247\n",
      "  Majority codon (from pileup): TCT\n",
      "  Majority codon count: 247\n",
      "  Number of unique codons discovered (including indels): 15\n",
      "\n",
      "  Total codon count (excluding indels): 256\n",
      "    Unique types: 5\n",
      "    Non-synonymous: 2\n",
      "\n",
      "    Highest frequency non-synonymous: TTT = F\n",
      "                        at frequency: 0.01562\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 43\n",
      "Discovered...\n",
      "  Total codon count (including indels): 378\n",
      "  Reference codon (from 3D7): AAT\n",
      "  Reference codon count: 290\n",
      "  Majority codon (from pileup): AAT\n",
      "  Majority codon count: 290\n",
      "  Number of unique codons discovered (including indels): 29\n",
      "\n",
      "  Total codon count (excluding indels): 313\n",
      "    Unique types: 11\n",
      "    Non-synonymous: 9\n",
      "\n",
      "    Highest frequency non-synonymous: GAT = D\n",
      "                        at frequency: 0.02236\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 44\n",
      "Discovered...\n",
      "  Total codon count (including indels): 371\n",
      "  Reference codon (from 3D7): TCA\n",
      "  Reference codon count: 294\n",
      "  Majority codon (from pileup): TCA\n",
      "  Majority codon count: 294\n",
      "  Number of unique codons discovered (including indels): 16\n",
      "\n",
      "  Total codon count (excluding indels): 308\n",
      "    Unique types: 8\n",
      "    Non-synonymous: 4\n",
      "\n",
      "    Highest frequency non-synonymous: CCA = P\n",
      "                        at frequency: 0.00974\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 45\n",
      "Discovered...\n",
      "  Total codon count (including indels): 371\n",
      "  Reference codon (from 3D7): TTT\n",
      "  Reference codon count: 307\n",
      "  Majority codon (from pileup): TTT\n",
      "  Majority codon count: 307\n",
      "  Number of unique codons discovered (including indels): 15\n",
      "\n",
      "  Total codon count (excluding indels): 313\n",
      "    Unique types: 4\n",
      "    Non-synonymous: 3\n",
      "\n",
      "    Highest frequency non-synonymous: TCT = S\n",
      "                        at frequency: 0.00639\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 46\n",
      "Discovered...\n",
      "  Total codon count (including indels): 372\n",
      "  Reference codon (from 3D7): ATG\n",
      "  Reference codon count: 271\n",
      "  Majority codon (from pileup): ATG\n",
      "  Majority codon count: 271\n",
      "  Number of unique codons discovered (including indels): 24\n",
      "\n",
      "  Total codon count (excluding indels): 297\n",
      "    Unique types: 9\n",
      "    Non-synonymous: 8\n",
      "\n",
      "    Highest frequency non-synonymous: ATA = I\n",
      "                        at frequency: 0.03704\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 47\n",
      "Discovered...\n",
      "  Total codon count (including indels): 374\n",
      "  Reference codon (from 3D7): AAT\n",
      "  Reference codon count: 306\n",
      "  Majority codon (from pileup): AAT\n",
      "  Majority codon count: 306\n",
      "  Number of unique codons discovered (including indels): 19\n",
      "\n",
      "  Total codon count (excluding indels): 320\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 5\n",
      "\n",
      "    Highest frequency non-synonymous: GAT = D\n",
      "                        at frequency: 0.01875\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 48\n",
      "Discovered...\n",
      "  Total codon count (including indels): 375\n",
      "  Reference codon (from 3D7): CTA\n",
      "  Reference codon count: 296\n",
      "  Majority codon (from pileup): CTA\n",
      "  Majority codon count: 296\n",
      "  Number of unique codons discovered (including indels): 17\n",
      "\n",
      "  Total codon count (excluding indels): 313\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 3\n",
      "\n",
      "    Highest frequency non-synonymous: CCA = P\n",
      "                        at frequency: 0.01278\n",
      "\n",
      "--------------------------------------------------------------------------------\n",
      "Checking Codon: 49\n",
      "Discovered...\n",
      "  Total codon count (including indels): 368\n",
      "  Reference codon (from 3D7): ACT\n",
      "  Reference codon count: 288\n",
      "  Majority codon (from pileup): ACT\n",
      "  Majority codon count: 288\n",
      "  Number of unique codons discovered (including indels): 17\n",
      "\n",
      "  Total codon count (excluding indels): 314\n",
      "    Unique types: 7\n",
      "    Non-synonymous: 4\n",
      "\n",
      "    Highest frequency non-synonymous: GCT = A\n",
      "                        at frequency: 0.02229\n",
      "\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'6'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-24-4fabcced0aaa>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgene_dt\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"strand\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'reverse'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m             \u001b[0mref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcomplement_map\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m             \u001b[0mprocessed_pileup\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcomplement_map\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mbase\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mbase\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mprocessed_pileup\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m         \u001b[0mcodon_ref_nts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m         \u001b[0mcodon_pileup\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessed_pileup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-24-4fabcced0aaa>\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mgene_dt\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"strand\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'reverse'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m             \u001b[0mref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcomplement_map\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m             \u001b[0mprocessed_pileup\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcomplement_map\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mbase\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mbase\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mprocessed_pileup\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m         \u001b[0mcodon_ref_nts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mref\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m         \u001b[0mcodon_pileup\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessed_pileup\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: '6'"
     ]
    }
   ],
   "source": [
    "with open(pileup_path, \"r\") as fn:\n",
    "    \n",
    "    # want to loop until you have a full codon\n",
    "    nts = 0\n",
    "    codon_position = 0\n",
    "    codon_ref_nts = []\n",
    "    codon_pileup = []\n",
    "    for i, line in enumerate(fn):\n",
    "        nts += 1\n",
    "        chrom, pos, ref, coverage, pileup, _ = line.split(\"\\t\")\n",
    "        processed_pileup = process_pileup(pileup, ref)\n",
    "        if gene_dt[\"strand\"] == 'reverse':\n",
    "            ref = complement_map[ref]\n",
    "            processed_pileup = \"\".join([complement_map[base] for base in processed_pileup])\n",
    "        codon_ref_nts.append(ref)\n",
    "        codon_pileup.append(processed_pileup)\n",
    "\n",
    "        if nts == 3:\n",
    "            codon_position += 1\n",
    "            print(\"--------------------------------------------------------------------------------\")\n",
    "            print(\"Checking Codon:\", codon_position)\n",
    "            # Search the codon for non-synonymous mutations above a specified frequency\n",
    "            assert len(codon_ref_nts) == 3\n",
    "            assert len(codon_pileup) == 3\n",
    "            \n",
    "            # Pre-process from lists into strings\n",
    "            codon_ref = \"\".join(codon_ref_nts)\n",
    "            amino_ref = codon_to_amino(codon_ref, genetic_code)\n",
    "            codon_pileup = [\"\".join(c) for c in zip(*codon_pileup)]\n",
    "            \n",
    "            # Get frequencies of codons (i.e. nucleotide level)\n",
    "            codon_frequencies = Counter(codon_pileup)\n",
    "            major_codon, major_codon_count = codon_frequencies.most_common(1)[0]\n",
    "            ref_codon_count = codon_frequencies[codon_ref]\n",
    "            total_codon_count = sum(codon_frequencies.values())\n",
    "\n",
    "            print(\"Discovered...\")\n",
    "            print(\"  Total codon count (including indels):\", total_codon_count)\n",
    "            print(\"  Reference codon (from 3D7):\", codon_ref)\n",
    "            print(\"  Reference codon count:\", ref_codon_count)\n",
    "            print(\"  Majority codon (from pileup):\", major_codon)\n",
    "            print(\"  Majority codon count:\", major_codon_count)\n",
    "            print(\"  Number of unique codons discovered (including indels):\", len(codon_frequencies))\n",
    "            print(\"\")\n",
    "\n",
    "            # Restrict to codon observations without indels\n",
    "            # - these can be used to look for non-synonymous changes\n",
    "            codon_noindel_frequencies = Counter(dict([(c, v) for c, v in codon_frequencies.items() \n",
    "                                                      if not \"+\" in c and not \"-\" in c]))\n",
    "            total_noindel_count = sum(codon_noindel_frequencies.values())\n",
    "\n",
    "            # Now get non-synonymous frequencies\n",
    "            codon_nonsyn_frequencies = Counter(dict([(c, v) for c, v in codon_noindel_frequencies.items()\n",
    "                                                     if codon_to_amino(c, genetic_code) != amino_ref]))\n",
    "            \n",
    "            \n",
    "            \n",
    "            \n",
    "            print(\"  Total codon count (excluding indels):\", total_noindel_count)\n",
    "            print(\"    Unique types:\", len(codon_noindel_frequencies))\n",
    "            print(\"    Non-synonymous:\", len(codon_nonsyn_frequencies))\n",
    "            print(\"\")\n",
    "            if len(codon_nonsyn_frequencies) > 0:\n",
    "                highest_nonsyn_codon, highest_nonsyn_count = codon_nonsyn_frequencies.most_common(1)[0]\n",
    "                print(\"    Highest frequency non-synonymous: %s = %s\" \n",
    "                      % (highest_nonsyn_codon, codon_to_amino(highest_nonsyn_codon, genetic_code)))\n",
    "                print(\"                        at frequency: %.05f\" \n",
    "                      % (highest_nonsyn_count/total_noindel_count))\n",
    "                \n",
    "                # Finally, check if any of the non-synonymous mutations are above\n",
    "                # set threshold\n",
    "                above = Counter(dict([(c, v) for c, v in codon_nonsyn_frequencies.items() \n",
    "                      if v/total_noindel_count > min_freq]))\n",
    "                if len(above) > 0:\n",
    "                    n_above = len(above)\n",
    "                    print(\"Detected %d non-synonymous mutations above frequency threshold of %.02f.\"\n",
    "                          % (n_above, min_freq))\n",
    "                    print(\"Storing results.\")\n",
    "                    \n",
    "                    for mut_codon, mut_counts in above.items():\n",
    "                        mutation_dt[\"position\"].append(codon_position)\n",
    "                        mutation_dt[\"total_counts\"].append(total_codon_count)\n",
    "                        mutation_dt[\"total_counts_noindel\"].append(total_noindel_count)\n",
    "                        \n",
    "                        mutation_dt[\"ref_codon\"].append(codon_ref)\n",
    "                        mutation_dt[\"ref_amino\"].append(amino_ref)\n",
    "                        mutation_dt[\"ref_freq\"].append(ref_codon_count/total_codon_count)\n",
    "                        \n",
    "                        mutation_dt[\"major_codon\"].append(major_codon)\n",
    "                        mutation_dt[\"major_amino\"].append(codon_to_amino(major_codon, genetic_code))\n",
    "                        mutation_dt[\"major_freq\"].append(major_codon_count/total_noindel_count)\n",
    "                        \n",
    "                        mutation_dt[\"n_mutation_types\"].append(len(codon_nonsyn_frequencies))\n",
    "                        mutation_dt[\"mutation_codon\"].append(mut_codon)\n",
    "                        mutation_dt[\"mutation_amino\"].append(codon_to_amino(mut_codon, genetic_code))\n",
    "                        mutation_dt[\"mutation_freq\"].append(mut_counts/total_noindel_count)\n",
    "\n",
    "            else:\n",
    "                print(\"    No non-synonymous mutations discovered.\")\n",
    "            print(\"\")\n",
    "            \n",
    "            # Once mutation search is complete, reset codon and continue\n",
    "            nts = 0\n",
    "            codon_ref_nts = []\n",
    "            codon_pileup = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "',.,,,..-1C.-1C,.......*..........-1C...-6CTAGTT....,,,,,,,,,,,......-1C...........*...,,,,,,,,,,,,*,,,,,,,,,,,,,........................*..*......**........,,,,,,,,,,,,,,,,,,,,,.........,,,,,,,............T........-1C.*......*....+1C.......G.....-1C.............,,-106ctagttagattcataaatgaattagaatcattctcgctactacttccgcttttatcatcactattgctgttaccaccagattccctatcatacgtcatagaaaaat,,,,,,,,....T,..-1C..,,....*.-1C....,,,..-1C,,,...,,,,,,,,.,..,,,.,,-1c,,,,,,..,,,-2ct,,.*.,,,,,,...,,*,,,,*,.,*,.,,..,,.,,*,,,,.-1C,,,-1c,T,,.,,.,*,,,,,.,'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pileup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_df = pd.DataFrame(mutation_dt)\n",
    "columns = [\"position\", \"total_counts\", \"total_counts_noindel\", \n",
    "           \"ref_codon\", \"ref_amino\", \"ref_freq\",\n",
    "           \"major_codon\", \"major_amino\", \"major_freq\",\n",
    "           \"n_mutation_types\",\n",
    "           \"mutation_codon\", \"mutation_amino\", \"mutation_freq\"]\n",
    "mutation_df = mutation_df[columns]\n",
    "print(\"--------------------------------------------------------------------------------\")\n",
    "print(\"Mutation scan complete.\")\n",
    "print(\"================================================================================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
